### PR TITLE
[Balance] Ability-ignoring effects no longer ignore the source's own Abilities

### PR DESCRIPTION
**Pull request by innerthunder**
## What are the changes?

This change makes it so that if a Pokemon has an ability or uses a move that ignores ability effects, those ability-ignoring effects do not apply to the user's own abilities. For example, if a Pokemon has Mold Breaker as an ability and Contrary as a passive ability, it using Overheat would now raise the user's Sp. Atk instead of ignoring Contrary and lowering it.

## Why am I doing these changes?

This change was requested by the Balance team, fitting in line with previous changes to Neutralizing Gas (#1662 ). Also, this interaction has been [reported as a bug](https://discord.com/channels/1125469663833370665/1267021266271866890) before in the Discord, which indicates that players may find this interaction unintuitive as it stands.

## What did change?

- `field/arena`: New field `Arena.ignoreAbilitySource` tracks the source of active ability-ignoring effects by battler index, and is set within `setIgnoreAbilities`.
- `field/pokemon`: `Pokemon.canApplyAbility` now checks that `arena.ignoreAbilitySource` does not match the Pokemon's battler index before ignoring abilities' effects.
- `phases`: Small adjustments reflecting the changes to `Arena.setIgnoreAbilities()`

### Screenshots/Videos

Clanging Scales from a Pokemon with Mold Breaker + Contrary after the changes:

https://github.com/user-attachments/assets/49e5dbd3-54a8-4666-8fd8-b236902ebc6b

## How to test the changes?

tested via overrides

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/field/arena.ts [hunk 1]
```diff
@@ -536,8 +537,13 @@ export class Arena {
...
   }
 
-  setIgnoreAbilities(ignoreAbilities: boolean = true): void {
+  setIgnoreAbilities(ignoreAbilities: boolean, ignoreAbilitySource: BattlerIndex | null = null): void {
     this.ignoreAbilities = ignoreAbilities;
+    if (ignoreAbilities) {
+      this.ignoreAbilitySource = ignoreAbilitySource;
+    } else {
+      this.ignoreAbilitySource = null;
+    }
```
Comment (DayKev) [inline src/field/arena.ts:546]:
```suggestion
    this.ignoreAbilitySource = ignoreAbilities ? ignoreAbilitySource : null;
```

**Comment by Xavion3** [comment]
It's unintuitive partially because Neutralising gas is bugged to not negate abilities like it should which creates an inconsistency, we can easily fix that by just reverting the change which broke it.

Also this is not a quality of life change, this is a significant change in how an ability works with significant balance and gameplay ramifications. This is even a nerf under some circumstances.

The bug report was also thrown off by the text of mold breaker not lining up with the effect, but this change does not solve that problem so that's not really a good excuse either. Unless we want to make it not ignore stuff like lightning rod.

**Review by Xavion3 — DISMISSED** [review]
As above, this currently fails to actually solve the problem laid out in the reasons for this PR.

File: src/field/arena.ts [hunk 2]
```diff
@@ -536,8 +537,9 @@ export class Arena {
     }
   }
 
-  setIgnoreAbilities(ignoreAbilities: boolean = true): void {
+  setIgnoreAbilities(ignoreAbilities: boolean, ignoreAbilitySource: BattlerIndex | null = null): void {
```
Comment (schmidtc1) [inline src/field/arena.ts:540]: To me, `ignoreAbilitySource` reads like a `boolean`. Maybe change the name to something like `abilityBattlerIndex`, `abilitySourceBattlerIndex`, etc.

Comment (Fontbane) [reply src/field/arena.ts:540]: Or `ignore**d**AbilitySource`

**Comment by DayKev** [comment]
> It's unintuitive partially because Neutralising gas is bugged to not negate abilities like it should which creates an inconsistency, we can easily fix that by just reverting the change which broke it.
> 
> Also this is not a quality of life change, this is a significant change in how an ability works with significant balance and gameplay ramifications. This is even a nerf under some circumstances.
> 
> The bug report was also thrown off by the text of mold breaker not lining up with the effect, but this change does not solve that problem so that's not really a good excuse either. Unless we want to make it not ignore stuff like lightning rod.

So we need balance team comments then.

**Comment by innerthunder** [comment]
Closing this as there is at least some evidence (via Sunsteel Strike/Moongeist Beam) that ability-ignoring effects apply to the user's ability in mainline; it's just that these effects are never available alongside ignorable abilities (hence Shadow Shield and Full Metal Body not being ignorable).

I may reopen this if the Game Design/Balance teams decide custom implementation is needed to remove potential anti-synergy between a Pokemon's passive and ability, but as it stands I'd say the current implementation is more accurate to mainline.

File: src/field/arena.ts (hunk 2 above)
Comment (innerthunder) [reply src/field/arena.ts:540]: maybe something along the lines of `attackerIndex`? It's not necessarily an ability that creates this effect (see Moongeist Beam/Sunsteel Strike), and `ignoredAbilitySource` kind of sounds like it's referring to the Pokemon whose ability is ignored, not the one who's doing the ignoring.

**Review by damocleas — APPROVED** [review]

**Review by DayKev — APPROVED** [review]

**Review by flx-sta — APPROVED** [review]

**Pull request merged**